core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -2352,6 +2354,35 @@ public static int subtractMonths(long t0, long t1) {
     return x;
   }
 
+  /**
+   * Implements the {@code .} (field access) operator on an object whose type is not known until
+   * runtime.
+   *
+   * A struct object can be represented in various ways by the runtime and depends on the
+   * {@link org.apache.calcite.adapter.enumerable.JavaRowFormat}.
+   */
+  public static Object structAccess(Object structObject, int index) {
+    if (structObject == null) {
+      return null;
+    }
+
+    if (structObject instanceof Object[]) {
+      return ((Object[]) structObject)[index];
+    } else if (structObject instanceof List) {
+      return ((List) structObject).get(index);
+    } else if (structObject instanceof Row) {
+      return ((Row) structObject).getObject(index);
+    } else {
+      Class<?> beanClass = structObject.getClass();
+      Field structField = beanClass.getDeclaredFields()[index];
```

Comment:
@zabetak ,
1) `getDeclaredFeilds()` does not guarantee the order of returned fields.
2) `getDeclaredFields()` per access is bad performance-wise